Debounce extensions query

tjkandala opened the pull request:
On webapp initialization, settings are updated in rapid succession. We fetch enabled extensions on each settings update, so this results in network console spam. Debounce settings changes to reduce spam.

eseliger commented on client/shared/src/extensions/helpers.ts:
```diff
@@ -18,6 +18,7 @@ export function viewerConfiguredExtensions({
     return from(settings).pipe(
         map(settings => extensionIDsFromSettings(settings)),
         distinctUntilChanged((a, b) => isEqual(a, b)),
+        debounceTime(200),
```
`debounceTime` also delays the emission, if there aren't any successors in a short time, meaning this would delay it by 200ms, even if no second request would have been triggered (see https://rxmarbles.com/#debounceTime `#1`)

I think what we would want here is only debouncing when at least 1 request was triggered in the last 200ms(?)

tjkandala replied:
Good catch! But wouldn't that lead to two requests on initialization -> logspam?

`---|--|--|--|-----`
`---|-----------|-`

This observable is subscribed to only by the extensions service in order to fetch and activate extensions, and extension enablement settings are usually changed through extension registry toggles, so I don't think we'd lose any UI responsiveness from debouncing here

eseliger replied:
Hmm yeah you're right that would fire twice.. 🤔

Yeah, I don't have too much worry about the toggle. I was only worried about the initial fetch, which is delayed by 200ms then and would defer startup by 200ms, no? Maybe I'm just not deep enough into the topic here, though 😬 

Do you happen to know what's the reason why the settings keep getting updated so often on app startup? 🤔 Just trying to understand whats happening here

felixfbecker commented:
While this treats the symptom at this specific location, why are the settings updated on the client in so rapid succession (or at all)? A lot of places listen to settings updates so are probably doing unnecessary work too

tjkandala commented:
@eseliger @felixfbecker I've found the source of the bug:
Settings are not emitted more than once on startup (`helpers.ts` is where `viewerConfiguredExtensions` lives)
![settings update logs](https://user-images.githubusercontent.com/37420160/108115114-ad844000-7067-11eb-9b40-2292ecb2258d.png)

The first subscriber to `viewerConfiguredExtensions` is the [contribution registry](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/shared/src/extensions/controller.ts#L159:20) by way of `ExtensionsService#activeExtensions`. 

The contribution registry combines all contribution entries and unsubscribes from existing contributions when new entries are added. So when, for instance, [hover actions are registered](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/shared/src/hover/actions.ts#L337:10), the request for configured extension is cancelled and immediately made again.

Shouldn't `viewerConfiguredExtensions` be multicast with a reference counter? The other subscriber, `ClientExtensions`, isn't instantiated [until the extension host worker is initialized](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/shared/src/api/client/connection.ts#L74-141), so `viewerConfiguredExtensions`' reference count drops to zero whenever new contributions from the webapp are registered and existing contributions observables are unsubscribed from.

eseliger commented:
Oh great investigation, thanks for digging deeper here!

mrnugget commented:
(Asking because I'm highly interested in seeing this get solved)

@tjkandala your analysis (which is really nice btw) sounds like the fix _right now_ is not sufficient, is that correct? I'm asking because this PR has been moved to "Review" in the board and I'm not sure if that means "ready for code review" or "ready to discuss this issue again"

felixfbecker commented:
@tjkandala what you proposed (multicasting) seems sound. I assume it would just need a replay (with buffer 1) then to keep the behavior the same.

tjkandala commented:
@felixfbecker It was multicast before, but since we counted references (subscribers), it would repeatedly cancel and restart the query since the query hadn't completed yet and couldn't be buffered. I removed `publishReplay` + `refCount` in favor of `shareReplay(1)`. This isn't "correct", since `viewerConfiguredExtensions` will now never close until `settings` is complete, but in reality this doesn't matter at all, since it's only unsubscribed from when `SourcegraphWebApp` unmounts. 

I worked a bit on the root cause, contribution registry unsubscribing from existing entries, but the current code solves the problem as well as we need it to (no delay added to extension activation). Extension contributions are the only `Observable` entry right now, so refactoring now wouldn't have any impact. I left TODO comments with context for when we have time, though.

@mrnugget The fix when I moved it to review was a band-aid, and so is this fix, but since it won't have a negative impact on users, and it'll improve developer experience/you're excited to see it merged :), I'm looking for review in the PRs current state
